# Web Scraping Lab

You will find in this notebook some web scraping exercises to practice your scraping skills using `requests` and `Beautiful Soup`.

**Tips:**

- Check the [response status code](https://http.cat/) for each request to ensure you have obtained the intended content.
- Look at the HTML code in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.
- Check out the css selectors.

### Useful Resources
- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### First of all, gathering our tools.

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import json

⚠️ **Again, please remember to limit your output before submission so that your code doesn't get lost in the output.**

#### Challenge 1 - Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [3]:
response1 = requests.get(url)
response1.status_code

soup1 = BeautifulSoup(response1.content)

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below (with different names):

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

In [4]:
developers = soup1.find_all('h1', attrs={'class':'h3 lh-condensed'})
developers[0]

name_developers = []

In [5]:
for dep in developers:
    name_developers.append(dep.text.strip())
name_developers[:5]

['Michele Locati', 'Han Xiao', 'Alon Zakai', 'Ha Thach', 'Joel Arvidsson']

####  Challenge 2 - Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [6]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [7]:
response2 = requests.get(url)
response2.status_code
soup2 = BeautifulSoup(response2.content) 


In [8]:
repositories = soup2.find_all('h1', attrs={'class':'h3 lh-condensed'})
repositories[0]

<h1 class="h3 lh-condensed">
<a data-hydro-click='{"event_type":"explore.click","payload":{"click_context":"TRENDING_REPOSITORIES_PAGE","click_target":"REPOSITORY","click_visual_representation":"REPOSITORY_NAME_HEADING","actor_id":null,"record_id":240315046,"originating_url":"https://github.com/trending/python?since=daily","user_id":null}}' data-hydro-click-hmac="4e4ae9fe56c93d4b1c3c4d96b377380b68f4727741403591b8b5bc5ee5cc0575" href="/jina-ai/jina">
<svg aria-hidden="true" class="octicon octicon-repo mr-1 text-gray" height="16" version="1.1" viewbox="0 0 16 16" width="16"><path d="M2 2.5A2.5 2.5 0 014.5 0h8.75a.75.75 0 01.75.75v12.5a.75.75 0 01-.75.75h-2.5a.75.75 0 110-1.5h1.75v-2h-8a1 1 0 00-.714 1.7.75.75 0 01-1.072 1.05A2.495 2.495 0 012 11.5v-9zm10.5-1V9h-8c-.356 0-.694.074-1 .208V2.5a1 1 0 011-1h8zM5 12.25v3.25a.25.25 0 00.4.2l1.45-1.087a.25.25 0 01.3 0L8.6 15.7a.25.25 0 00.4-.2v-3.25a.25.25 0 00-.25-.25h-3.5a.25.25 0 00-.25.25z" fill-rule="evenodd"></path></svg>
<span class="text

In [63]:
for repo in repositories:
    name_repositories.append(repo.text.strip())


  

In [78]:
name_repositories2 = []
for x in name_repositories:
    name_repositories2.append(x.replace('/\n\n ', ''))



In [79]:
name_repositories2

['jina-ai      jina',
 'MTK-bypass      bypass_utility',
 'ansible      awx',
 'Asabeneh      30-Days-Of-Python',
 'PaddlePaddle      PaddleOCR',
 'RyanElliott10      wsbtickerbot',
 'facebookresearch      detectron2',
 'public-apis      public-apis',
 'huggingface      transformers',
 'OWASP      CheatSheetSeries',
 'demisto      content',
 'pandas-dev      pandas',
 'facebookresearch      ParlAI',
 'facebookresearch      detr',
 'pytorch      vision',
 'wagtail      wagtail',
 'great-expectations      great_expectations',
 'sanic-org      sanic',
 'pytorch      fairseq',
 'mingrammer      diagrams',
 'nodejs      node-gyp',
 'deepfakes      faceswap',
 'Vonng      ddia',
 'spulec      moto',
 'samuelcolvin      pydantic',
 'jina-ai      jina',
 'MTK-bypass      bypass_utility',
 'ansible      awx',
 'Asabeneh      30-Days-Of-Python',
 'PaddlePaddle      PaddleOCR',
 'RyanElliott10      wsbtickerbot',
 'facebookresearch      detectron2',
 'public-apis      public-apis',
 'huggingface 

#### Challenge 3 - Display all the image links from Walt Disney wikipedia page

In [50]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [51]:
disney = requests.get(url)
disney.status_code
disney = BeautifulSoup(disney.content)

In [52]:
image = disney.find_all("a", class_="image")


In [53]:
image_list = []
for ima in image:
    image_list.append(ima.get("href"))
image_list[:5]

['/wiki/File:Walt_Disney_1946.JPG',
 '/wiki/File:Walt_Disney_1942_signature.svg',
 '/wiki/File:Walt_Disney_envelope_ca._1921.jpg',
 '/wiki/File:Trolley_Troubles_poster.jpg',
 '/wiki/File:Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg']

#### Challenge 4 - Retrieve all links to pages on Wikipedia that refer to some kind of Python.

In [54]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 

In [55]:
python =requests.get(url)
python.status_code
python = BeautifulSoup(python.content)

In [56]:
link = python.find_all("a", {"href": "/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)" })
link

[<a href="/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)" title="Python (Coney Island, Cincinnati, Ohio)">Python (Coney Island, Cincinnati, Ohio)</a>]

In [57]:
link_python = []
for x in link:
    link_python.append(x.get("href"))
    
link_python    

['/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)']

#### Challenge 5 - Number of Titles that have changed in the United States Code since its last release point 

In [18]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [19]:
title = requests.get(url)
title.status_code

titles = BeautifulSoup(title.content)

In [20]:
number = titles.find_all("div", attrs={"class":["usctitle", "usctitlechanged"]}) ## uscitem

#La longitud de los que hemos sacado es el número de títulos que tenemos
len(number)

56

In [ ]:
''' Intente sacar una función buscando el patrón del título pero no lo conseguí
import re
def number_titles (x):
    x = str(x)
    return len(re.findall(r'Title \d - \w', x))
'''

#### Challenge 6 - A Python list with the top ten FBI's Most Wanted names 

In [21]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [22]:
names = requests.get(url)
names.status_code
nam = BeautifulSoup(names.content, 'html.parser')

In [23]:
fbi = nam.find_all('h3', attrs={'class':'title'})
fbi[1]

most_wanted = []
for i in fbi:
    most_wanted.append(i.text.strip())

most_wanted

['ALEJANDRO ROSALES CASTILLO',
 'ARNOLDO JIMENEZ',
 'JASON DEREK BROWN',
 'ALEXIS FLORES',
 'JOSE RODOLFO VILLARREAL-HERNANDEZ',
 'EUGENE PALMER',
 'RAFAEL CARO-QUINTERO',
 'ROBERT WILLIAM FISHER',
 'BHADRESHKUMAR CHETANBHAI PATEL',
 'YASER ABDEL SAID']

#### Challenge 7 - List all language names and number of related articles in the order they appear in wikipedia.org

In [24]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'


In [25]:
wiki = requests.get(url)
wiki.status_code
lang_wiki = BeautifulSoup(wiki.content)

In [26]:
language = lang_wiki.find_all("div", class_="central-featured-lang")
language[1]

<div class="central-featured-lang lang2" dir="ltr" lang="ja">
<a class="link-box" data-slogan="フリー百科事典" href="//ja.wikipedia.org/" id="js-link-box-ja" title="Nihongo — ウィキペディア — フリー百科事典">
<strong>日本語</strong>
<small><bdi dir="ltr">1 250 000+</bdi> <span>記事</span></small>
</a>
</div>

In [ ]:
language_list = []
for wiki in language:
    language_list.append(wiki.get("title"))
language_list = [wiki.replace('\\xa0',' ') for wiki in language_list]    

#### Challenge 8 - A list with the different kind of datasets available in data.gov.uk

In [29]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [30]:
lists = requests.get(url)
lists.status_code
lists_uk = BeautifulSoup(lists.content)

In [31]:
dataset = lists_uk.find_all("h3", class_="govuk-heading-s dgu-topics__heading")
dataset[1]

<h3 class="govuk-heading-s dgu-topics__heading"><a class="govuk-link" href="/search?filters%5Btopic%5D=Crime+and+justice">Crime and justice</a></h3>

In [32]:
uk_datasets = []
for dat in dataset:
    uk_datasets.append(dat.text.strip())

uk_datasets

['Business and economy',
 'Crime and justice',
 'Defence',
 'Education',
 'Environment',
 'Government',
 'Government spending',
 'Health',
 'Mapping',
 'Society',
 'Towns and cities',
 'Transport']

#### Challenge 9 - Top 10 languages by number of native speakers stored in a Pandas Dataframe

In [33]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [34]:
top = requests.get(url)
top.status_code
top_lan = BeautifulSoup(top.content)

In [35]:
table = top_lan.find("table", {"class":"wikitable sortable"})
table

languages = table.find_all("a")
languages

#ahora vamos a sacar los paises:
countries = []
for i in languages:
    countries.append(i.get("title"))
    
countries[2:12]



['Mandarin Chinese',
 'Sino-Tibetan languages',
 'Varieties of Chinese',
 'Spanish language',
 'Indo-European languages',
 'Romance languages',
 'English language',
 'Indo-European languages',
 'Germanic languages',
 'Hindi']

### Stepping up the game

####  Challenge 10 - The 20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

#### Challenge 11 - IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [ ]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

#### Challenge 12 - Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [36]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

In [37]:
film = requests.get(url)
film.status_code
film = BeautifulSoup(film.content, 'html.parser')


In [38]:
import random
def titles(url):
    page = requests.get(url)
    film = BeautifulSoup(page.content, 'html.parser')
    movies = film.find_all("td", class_="titleColumn")
    movies = movies[:10]
        
    top_movies = []
    for i in movies:
        top_movies.append(i.text.strip())

    return top_movies
    

In [39]:
def years (url):
    page = requests.get(url)
    date = BeautifulSoup(page.content, 'html.parser')
    date = date.find_all("span", class_="secondaryInfo")
    date = date[:10]
    
    year = []
    for i in date:
        year.append(i.text.strip())
    return year

In [40]:
def califications (url):
    page = requests.get(url)
    cal = BeautifulSoup(page.content, 'html.parser')
    cal = cal.find_all("td", class_="ratingColumn imdbRating")
    cal = cal[:10]
    
    calification = []
    for i in cal:
        calification.append(i.text.strip())
    return calification


In [41]:
title = titles('http://www.imdb.com/chart/top')

In [42]:
year = years('http://www.imdb.com/chart/top')

In [43]:
calification = califications('http://www.imdb.com/chart/top')
calification

['9.2', '9.1', '9.0', '9.0', '8.9', '8.9', '8.9', '8.8', '8.8', '8.8']

In [44]:
df = pd.DataFrame()
df['Films']  = title
df['Year']  = year
df['Califications']  = calification
df


Films    Year Califications
0                  1.\n      Cadena perpetua\n(1994)  (1994)           9.2
1                       2.\n      El padrino\n(1972)  (1972)           9.1
2             3.\n      El padrino: Parte II\n(1974)  (1974)           9.0
3              4.\n      El caballero oscuro\n(2008)  (2008)           9.0
4            5.\n      12 hombres sin piedad\n(1957)  (1957)           8.9
5            6.\n      La lista de Schindler\n(1993)  (1993)           8.9
6  7.\n      El señor de los anillos: El retorno ...  (2003)           8.9
7                     8.\n      Pulp Fiction\n(1994)  (1994)           8.8
8       9.\n      El bueno, el feo y el malo\n(1966)  (1966)           8.8
9  10.\n      El señor de los anillos: La comunid...  (2001)           8.8

In [46]:
for column in df.columns:
    df["Films"] = df["Films"].replace('(\d)', '', regex=True)
    df["Films"] = df["Films"].replace('( \n)', '', regex=True)
df

Films    Year Califications
0                           .      Cadena perpetua()  (1994)           9.2
1                                .      El padrino()  (1972)           9.1
2                      .      El padrino: Parte II()  (1974)           9.0
3                       .      El caballero oscuro()  (2008)           9.0
4                       .       hombres sin piedad()  (1957)           8.9
5                     .      La lista de Schindler()  (1993)           8.9
6  .      El señor de los anillos: El retorno del...  (2003)           8.9
7                              .      Pulp Fiction()  (1994)           8.8
8                .      El bueno, el feo y el malo()  (1966)           8.8
9  .      El señor de los anillos: La comunidad d...  (2001)           8.8

#### Challenge 13 - Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [ ]:
def weather(city):
    pass

#### Challenge 14 - Book name,price and stock availability as a pandas dataframe.

In [ ]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

**Did you limit your output? Thank you! 🙂**